lib/sdf-server/src/server/service/graphviz.rs

```diff
@@ -246,10 +283,99 @@ pub async fn schema_variant(
     Ok(Json(GraphVizResponse {
         nodes,
         edges,
-        root_node_id,
+        root_node_id: Some(root_node_id),
     }))
 }
 
+pub async fn components(
+    HandlerContext(builder): HandlerContext,
+    AccessBuilder(request_ctx): AccessBuilder,
+    Query(request): Query<GraphVizRequest>,
+) -> GraphVizResult<Json<GraphVizResponse>> {
+    let ctx = builder.build(request_ctx.build(request.visibility)).await?;
+
+    let workspace_snapshot = ctx.workspace_snapshot()?;
+
+    let mut func_nodes = vec![];
+    let mut nodes = vec![];
+    let mut edges = vec![];
+    let mut added_nodes = HashSet::new();
+    let mut added_edges = HashSet::new();
+
+    let components = Component::list(&ctx).await?;
+
+    for component in &components {
+        if !added_nodes.contains(&component.id()) {
+            added_nodes.insert(component.id());
+        } else {
+            continue;
+        }
```

Comment:
I don't believe it is possible to see two components twice in the way that this is written.

Reply:
Thats what our first hunch was as well. Unless somehow there was a circular relationship between two components?

Reply:
@nickgerace It starts with a list of the components in the diagram. If we don't do this filtering here, depending on the order that the components vec is in, we could add the component because it has an incoming connection from another component, and then try to add the component again on its own loop. We saw this behavior while running the system